### PR TITLE
Improve CI/CD

### PR DESCRIPTION
**Pull request by tmattio**
This PR improves the CI CD by:

- Caching Opam dependencies
- Installing dependencies from lock files, which will make the builds more reproducible
- Splitting the steps and giving them names to identify failures quickly

### CODE REVIEW

**Review by rgrinberg — APPROVED** [review]

**Comment by anuragsoni** [comment]
@tmattio restoring cache might cause issues with a later switch creation. Right now the ocaml action uses `opam switch create "$1" "$1"` to create a switch. I think that'll fail if there's an existing switch with the same name already present.

https://github.com/avsm/setup-ocaml/blob/4b93f69dd17a2b0bd35732d2573e57e307085f1e/lib/install-ocaml-unix.sh#L11

**Comment by tmattio** [comment]
@anuragsoni That's what I noticed after merging. I'll comment out the caching part until there is a workaround. That's a shame, the CI went from taking 20min to 2min!

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+        with:
+          ocaml-version: ${{ matrix.ocaml-version }}
+
+      - name: Install dependencies
+        if: steps.cache-opam.outputs.cache-hit != 'true'
+        run: |
+          opam install . --deps-only --with-doc --with-test --locked --unlock-base
+          opam install ppx_sexp_conv base64 ppx_yojson_conv --skip-updates
+
+      - name: Upgrade dependencies
```
Comment (smorimoto) [inline .github/workflows/ci.yml:48]: This is NOT for upgrade. This step checks the consistency of the installed opam packages.

File: .github/workflows/format.yml [hunk 2]
```diff
@@ -1,21 +1,52 @@
...
+        uses: avsm/setup-ocaml@v1
+        with:
+          ocaml-version: ${{ matrix.ocaml-version }}
+
       - name: Install dependencies
-        run: opam install dune ocamlformat
+        if: steps.cache-opam.outputs.cache-hit != 'true'
+        run: opam install ocamlformat
+
+      - name: Upgrade dependencies
```
Comment (smorimoto) [inline .github/workflows/format.yml:43]: Same here.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (tmattio) [reply .github/workflows/ci.yml:48]: Thanks for the review @smorimoto! I made the changes in #184

**Pull request merged**